test/support/github_case/success_api.ex

```diff
@@ -0,0 +1,55 @@
+defmodule CodeCorps.GitHubCase.SuccessAPI do
+  @moduledoc ~S"""
+  A mocked github API layer which returns a default successful response for all
+  github api requests.
+
+  All tests in the test environment use this module as a mock for github api
+  requests by default.
+
+  If certain tests explicitly depend on the data returned by github, they can be
+  mocked individually using the `CodeCorps.GitHubCase.Helpers.with_mock_api`
+  macro.
+
+  As support for new github endpoints is added, defaults for those endpoints
+  should be added here.
+
+  To assert a request has been made to github as a result as an action, the
+  `assert_received` test helper can be used:
+
+  ```
+  assert_received({:get, "https://api.github.com/user", headers, body, options})
+  ```
+  """
+
+  import CodeCorps.GitHubCase.Helpers
+
+  def request(method, url, headers, body, options) do
+    send(self(), {method, url, headers, body, options})
+
+    {:ok, mock_response(method, url, headers, body, options)}
+  end
+
+  defp mock_response(:post, "https://github.com/login/oauth/access_token", _, _, _) do
+    %{"access_token" => "foo_auth_token"}
+  end
+  defp mock_response(method, "https://api.github.com/" <> endpoint, headers, body, options) do
+    mock_api_response(method, endpoint |> String.split("/"), headers, body, options)
```

Comment:
There can be dynamic parts to a request url, so this is how we deal with those. We can't pattern match a string when a dynamic part is not the last part of the string, so we split it into a list of parts and pattern match the list in a _subfunction_